### PR TITLE
feat: allow function allowList and keyGenerator to be async/await

### PR DESCRIPTION
**Pull request by DiogoMarques2003**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by DiogoMarques2003** [comment]
I think isn't need add new unit test, but if is necessary pls tellme

File: index.d.ts [hunk 1]
```diff
@@ -94,12 +94,12 @@ export interface RateLimitOptions {
    * @deprecated Use `allowList` property
    */
   whitelist?: string[] | ((req: FastifyRequest, key: string) => boolean);
-  allowList?: string[] | ((req: FastifyRequest, key: string) => boolean);
+  async allowList?: string[] | ((req: FastifyRequest, key: string) => boolean);
```
Comment (Uzlopak) [inline index.d.ts:97]:
```suggestion
  allowList?: string[] | ((req: FastifyRequest, key: string) => Promise<boolean>);
```

File: index.d.ts [hunk 2]
```diff
@@ -94,12 +94,12 @@ export interface RateLimitOptions {
...
   continueExceeding?: boolean;
   skipOnError?: boolean;
   ban?: number;
   onBanReach?: (req: FastifyRequest, key: string) => void;
-  keyGenerator?: (req: FastifyRequest) => string | number;
+  async keyGenerator?: (req: FastifyRequest) => string | number;
```
Comment (Uzlopak) [inline index.d.ts:102]:
```suggestion
  keyGenerator?: (req: FastifyRequest) => string | number | Promise<string | number>;
```

**Comment by Uzlopak** [comment]
It seems that your code is broken.

File: index.d.ts [hunk 3]
```diff
@@ -94,12 +94,12 @@ export interface RateLimitOptions {
    * @deprecated Use `allowList` property
    */
   whitelist?: string[] | ((req: FastifyRequest, key: string) => boolean);
-  allowList?: string[] | ((req: FastifyRequest, key: string) => boolean);
+  allowList?: string[] | ((req: FastifyRequest, key: string) => Promise<boolean>);
```
Comment (Uzlopak) [inline index.d.ts:97]:
```suggestion
  allowList?: string[] | ((req: FastifyRequest, key: string) => boolean | Promise<boolean>);
```

File: index.js [hunk 4]
```diff
@@ -160,7 +160,7 @@ async function rateLimitPlugin (fastify, settings) {
 
 async function buildRouteRate (pluginComponent, params, routeOptions) {
   const hook = params.hook || defaultHook
-  const hookHandler = rateLimitRequestHandler(params, pluginComponent)
+  const hookHandler = await rateLimitRequestHandler(params, pluginComponent)
```
Comment (Uzlopak) [inline index.js:163]:
```suggestion
  const hookHandler = rateLimitRequestHandler(params, pluginComponent)
```

File: index.js [hunk 5]
```diff
@@ -170,7 +170,7 @@ async function buildRouteRate (pluginComponent, params, routeOptions) {
   }
 }
 
-function rateLimitRequestHandler (params, pluginComponent) {
+async function rateLimitRequestHandler (params, pluginComponent) {
```
Comment (Uzlopak) [inline index.js:173]:
```suggestion
function rateLimitRequestHandler (params, pluginComponent) {
```

File: index.js [hunk 6]
```diff
@@ -274,7 +274,7 @@ function defaultErrorResponse (req, context) {
   return err
 }
 
-function defaultOnBanReach (req, key) {}
+function defaultOnBanReach (req, key) { }
```
Comment (Uzlopak) [inline index.js:277]:
```suggestion
function defaultOnBanReach (req, key) {}
```

**Comment by Uzlopak** [comment]
@DiogoMarques2003 
Please provide some unit tests.

**Comment by DiogoMarques2003** [comment]
But @mcollina i need consult the database in this functions, and the preValidation is executed after the ratelimit system

**Comment by Uzlopak** [comment]
onRequestRateLimiter is already async. 

If you think that the await in front of the two potential async functions is a performance bottleneck, we could detect if one of those methods is async like this

```js
const asyncFunctionPrototype = Object.getPrototypeOf(async function() {});

function isAsyncFunction(v) {
  return (
    typeof v === 'function' &&
    Object.getPrototypeOf(v) === asyncFunctionPrototype
  );
};
```

and if so have a implementation were we have await and another one were we dont await. 

But again, onRequestRateLimiter is async.

**Comment by DiogoMarques2003** [comment]
@Uzlopak i need to do this on the allowList and keyGenerator functions:
image
image

can i do this databases requests in other place ?
I think no, and because of that it opens in this pull

**Comment by Uzlopak** [comment]
I was answering mcollina

The only reason to reject this PR would be for me, when the performance decreases. And awaiting a non async function decreases the performance. So if the worry of mcollina is that we decrease the perf. by unnecessary awaiting then we can check if we either have an async function and then use it with await or sync methods and dont await. But this means some refactoring to avoid simply duplicating the current function.

**Comment by DiogoMarques2003** [comment]
oh ok ok

**Comment by Fdawgs** [comment]
> The only reason to reject this PR would be for me, when the performance decreases. And awaiting a non async function decreases the performance.

Got any reading material around this @Uzlopak? Is it that is awaiting a sync function causes the returned value become a resolved Promise object, and the Promise causes overhead?

**Comment by Uzlopak** [comment]
@Fdawgs 

I dont think that await wraps a sync method into a Promise.

I only had this issue once about 3 years ago: A sync function was awaited. When I removed the await from the sync-function the unit test became flaky. Investigating further, it revealed a very tight race condition in the sync-function. By awaiting the sync function it resulted in a short delay, which concealed the underlying race condition. I never investigated it further. 
I assume that await puts the result into the next run of the event loop, so the sync method was finished and did not show the race condition. But this is just an assumption.

**Comment by mcollina** [comment]
The problem is not about promises but rather as a conscious design decision when implementing this module: providing some limited protection against API misuse and DoS attacks.

If we need to check against a _database_ to load some data, you are executing a database query before the check, and therefore you are likely vulnerable. In other terms: looking those things up from a database is counterproductive. This PR makes it _simpler_ to do something insecure and I'm not convinced we should support it.

---

You can achieve what you want by adding an `'onRequest'` hook before registering this module. In the `'onRequest'` hook, you should add those values to the request object. __I do not recommend this__.

**Comment by DiogoMarques2003** [comment]
> The problem is not about promises but rather as a conscious design decision when implementing this module: providing some limited protection against API misuse and DoS attacks.
> 
> If we need to check against a _database_ to load some data, you are executing a database query before the check, and therefore you are likely vulnerable. In other terms: looking those things up from a database is counterproductive. This PR makes it _simpler_ to do something insecure and I'm not convinced we should support it.
> 
> You can achieve what you want by adding an `'onRequest'` hook before registering this module. In the `'onRequest'` hook, you should add those values to the request object. **I do not recommend this**.

Can you show an example to add the database requests in `'onRequest'` to be executed before calling the rate limit ?

**Comment by mcollina** [comment]
Something like this should work

```js
app.addHook('onRequest', async (req) => {
  // Don't do this, it's very unsafe
  const { id, allowList } = await callMyDb()
  req.aLongPropertyName = id
  req.aLongPropertyName2 = allowList
})
fastify.register(require('@fastify/rate-limit'), ...)
```

**Comment by DiogoMarques2003** [comment]
> The problem is not about promises but rather as a conscious design decision when implementing this module: providing some limited protection against API misuse and DoS attacks.
> 
> If we need to check against a _database_ to load some data, you are executing a database query before the check, and therefore you are likely vulnerable. In other terms: looking those things up from a database is counterproductive. This PR makes it _simpler_ to do something insecure and I'm not convinced we should support it.
> 
> You can achieve what you want by adding an `'onRequest'` hook before registering this module. In the `'onRequest'` hook, you should add those values to the request object. **I do not recommend this**.

The max function is async, so maybe can also be

**Comment by mcollina** [comment]
Ok, I'm sold. Let's make this possible.

**Comment by Uzlopak** [comment]
lol, I will give the approval so that you dont have to :D

**Review by Uzlopak — APPROVED** [review]

**Review by Eomm — CHANGES_REQUESTED** [review]
> I think isn't need add new unit test, but if is necessary pls tellme

Without a test we can't assure this feature will be removed accidentally from other PRs.

Please, add 2 tests (happy path and error case) for each change

**Comment by DiogoMarques2003** [comment]
image

@Eomm In my pc the all tests pass, why here in github don't pass ?

**Comment by Uzlopak** [comment]
Your tests fail in the CI pipeline.

**Comment by DiogoMarques2003** [comment]
> Your tests fail in the CI pipeline.

What should I do ?

**Comment by Eomm** [comment]
Could you try to rebase the branch?

```
git rebase upstream/master
```

**Comment by Uzlopak** [comment]
I rerun the workflow, but it seems that the tests are now flaky with the additional awaits.

**Comment by DiogoMarques2003** [comment]
> Reference in ne

this is strange because on my computer they all ran correctly

**Comment by DiogoMarques2003** [comment]
image

this is very stranger

**Comment by Eomm** [comment]
> > Reference in ne
> 
> this is strange because on my computer they all ran correctly

It may depend on the node.js version.

Here are the logs of the run:
https://github.com/fastify/fastify-rate-limit/runs/7904283030?check_suite_focus=true

It seems that the namespace test is always failing.

**Comment by DiogoMarques2003** [comment]
> > > Reference in ne
> > 
> > 
> > this is strange because on my computer they all ran correctly
> 
> It may depend on the node.js version.
> 
> Here are the logs of the run: https://github.com/fastify/fastify-rate-limit/runs/7904283030?check_suite_focus=true
> 
> It seems that the namespace test is always failing.

Maybe, i'm using version 16.6.0

**Comment by DiogoMarques2003** [comment]
@Eomm any soluction for this ?

**Comment by DiogoMarques2003** [comment]
> 

for some reason myFunction.constructor.name don't exists

**Comment by Eomm** [comment]
it was a code example, you need to use your own function `keyGenerator` and `allowList`

**Comment by DiogoMarques2003** [comment]
> it was a code example, you need to use your own function `keyGenerator` and `allowList`

yes, i do this but don't work in  any test 
image

**Comment by Eomm** [comment]
Should not it be `params.allowList?.constructor.name`?

**Comment by Uzlopak** [comment]
@Eomm 
Its a race condition. PR incoming

**Comment by Uzlopak** [comment]
#261 fixes the issues.

**Comment by Uzlopak** [comment]
You can actually use a prototype check

```js
const asyncFunctionPrototype = Object.getPrototypeOf(async function() {})

function isAsyncFunction(v) {
    return (
      typeof v === 'function' &&
      Object.getPrototypeOf(v) === asyncFunctionPrototype
    )
  }
```

**Comment by Uzlopak** [comment]
@Eomm 
What needs to be done to get this PR merged?

**Comment by Eomm** [comment]
- a quick doc note
- the isAsyncFunction check

**Review by Eomm — APPROVED** [review]

**Comment by Uzlopak** [comment]
@Eomm 
The isAsyncFunction-check is not added. I think adding that would be maybe not that smart, as a sync function returning a promise would not be detected.

**Comment by Eomm** [comment]
I'm ok landing it, there is a blocking review from @mcollina

**Comment by Uzlopak** [comment]
@mcollina 

PTAL

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**